plugins/opds.koplugin/opdsbrowser.lua

```diff
@@ -1095,46 +1346,296 @@ function OPDSBrowser:showDownloadListItemDialog(item)
-    end, info)
-    if completed then
-        UIManager:close(info)
-    end
-    local dl_count = #self.downloads
-    for i = dl_count, 1, -1 do
-        local item = self.downloads[i]
-        if downloaded and downloaded[item.file] then
-            table.remove(self.downloads, i)
-        else -- if subprocess has been interrupted, check for the downloaded file
+        local dl_count = 0
+        local dl_size = #dl_list
+        for i = dl_size, 1, -1 do
+            local item = dl_list[i]
+            if downloaded and downloaded[item.file] then
+                dl_count = dl_count + 1
+                table.remove(dl_list, i)
+            else -- if subprocess has been interrupted, check for the downloaded file
             local attr = lfs.attributes(item.file)
-            if attr then
-                if attr.size > 0 then
-                    table.remove(self.downloads, i)
-                else -- incomplete download
-                    os.remove(item.file)
+                if attr then
+                    if attr.size > 0 then
+                        table.remove(dl_list, i)
+                        if attr.modification > os.time() - 300 then -- Only count files touched in the last 5 mins
+                            dl_count = dl_count + 1
+                        end
+                    else -- incomplete download
+                        os.remove(item.file)
+                    end
                 end
             end
         end
+        local duplicate_count
+        if duplicate_list then
+            duplicate_count = #duplicate_list
+        else
+            duplicate_count = 0
+        end
+        dl_count = dl_count - duplicate_count
+        -- Make downloaded count timeout if there's a duplicate file prompt
+        local timeout = nil
+        if duplicate_count > 0 then
+            timeout = 3
+        end
+        if dl_count > 0 then
+            UIManager:show(InfoMessage:new{ text = T(N_("1 book downloaded", "%1 books downloaded", dl_count), dl_count), timeout = timeout,})
+        end
+        if not self.sync then
+            self:updateDownloadListItemTable()
+            self.download_list_updated = true
+            self._manager.updated = true
+        else
+            self.pending_syncs = dl_list
+            self._manager.updated = true
+        end
+        return duplicate_list
     end
-    dl_count = dl_count - #self.downloads
-    if dl_count > 0 then
-        self:updateDownloadListItemTable()
-        self.download_list_updated = true
-        self._manager.updated = true
-        UIManager:show(InfoMessage:new{ text = T(N_("1 book downloaded", "%1 books downloaded", dl_count), dl_count) })
+
+    local duplicate_list = dismissable_download()
+
+    if duplicate_list and #duplicate_list > 0 then
+        local textviewer
+        local text = _("These files are already on the device: \n")
+        for _, entry in ipairs(duplicate_list) do
+            text = text .. entry.file .. "\n"
```

Comment:
These massive concatenations are memory expensive.
We usually build an array of the text pieces and then use `table.concat(t, "\n")`.